FreeRTOS_Stream_Buffer.c

```diff
@@ -43,6 +43,226 @@
+ *        reaching the head.
+ * @param[in] pxBuffer: The circular stream buffer.
+ * @return The number of items which can be read from the tail before
+ *        reaching the head.
+ */
+size_t uxStreamBufferGetSize( const StreamBuffer_t * pxBuffer )
+{
+    size_t uxHead = pxBuffer->uxHead;
+    size_t uxTail = pxBuffer->uxTail;
+
+    return uxStreamBufferDistance( pxBuffer, uxTail, uxHead );
+}
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Get the space between the mid pointer and the head in the stream
+ *        buffer.
+ * @param[in] pxBuffer: The circular stream buffer.
+ * @return The space between the mid pointer and the head.
+ */
+size_t uxStreamBufferMidSpace( const StreamBuffer_t * pxBuffer )
+{
+    size_t uxHead = pxBuffer->uxHead;
+    size_t uxMid = pxBuffer->uxMid;
+
+    return uxStreamBufferDistance( pxBuffer, uxMid, uxHead );
+}
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Move Clear the stream buffer.
+ * @param[in] pxBuffer: The circular stream buffer.
+ */
+void vStreamBufferClear( StreamBuffer_t * pxBuffer )
+{
+    /* Make the circular buffer empty */
+    pxBuffer->uxHead = 0U;
+    pxBuffer->uxTail = 0U;
+    pxBuffer->uxFront = 0U;
+    pxBuffer->uxMid = 0U;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Move the mid pointer forward by given byte count
+ * @param[in] pxBuffer: The circular stream buffer.
+ * @param[in] uxCount: The byte count by which the mid pointer is to be moved.
+ */
+void vStreamBufferMoveMid( StreamBuffer_t * pxBuffer,
+                           size_t uxCount )
+{
+    /* Increment uxMid, but no further than uxHead */
+    size_t uxSize = uxStreamBufferMidSpace( pxBuffer );
+    size_t uxMid = pxBuffer->uxMid;
+    size_t uxMoveCount = uxCount;
+
+    if( uxMoveCount > uxSize )
+    {
+        uxMoveCount = uxSize;
+    }
+
+    uxMid += uxMoveCount;
+
+    if( uxMid >= pxBuffer->LENGTH )
+    {
+        uxMid -= pxBuffer->LENGTH;
+    }
+
+    pxBuffer->uxMid = uxMid;
+}
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Check whether the value in left is less than or equal to the
+ *        value in right from the perspective of the circular stream
+ *        buffer.
+ * @param[in] pxBuffer: The circular stream buffer.
+ * @param[in] uxLeft: The left pointer in the stream buffer.
+ * @param[in] uxRight: The right value pointer in the stream buffer.
+ * @return pdTRUE if uxLeft <= uxRight, else pdFALSE.
+ */
+BaseType_t xStreamBufferLessThenEqual( const StreamBuffer_t * pxBuffer,
+                                       const size_t uxLeft,
+                                       const size_t uxRight )
+{
+    BaseType_t xReturn;
+    size_t uxTail = pxBuffer->uxTail;
+
+    /* Returns true if ( uxLeft < uxRight ) */
+    if( ( ( ( uxLeft < uxTail ) ? 1U : 0U ) ^ ( ( uxRight < uxTail ) ? 1U : 0U ) ) != 0U )
```

Comment:
Is there a simpler way to express this?

Reply:
It use to be much simpler, but MISRA didn't like it because the XOR was applied to two boolean expressions:
~~~c
if( ( uxLeft < uxTail ) ^ ( uxRight < uxTail ) )
~~~
and also: XOR results in a number, not a boolean :-(

Reply:
maybe something like:
```c
xReturn = pdFalse;
if( ( uxLeft - uxTail ) <= ( uxRight - uxTail ) )
{
    xReturn = pdTRUE;
}
```

Reply:
That's great! 
Just to be sure, I tested it for all 1000 combinations under 10 and got the same results as with the former complicated formula. Thanks!